File: solver/state.py
from typing import Optional

from rules.rules import MIN_VALUE

from .types import GameMode, Grid
from .utils import diag_indexes
from .validation import (
    resolve_max_cell_value,
    validate_and_normalize_known_grid,
    validate_diagonal_partial_sums,
    validate_partial_sums,
)


InitialState = tuple[
    Grid,
    list[int],
    list[int],
    list[int],
    list[int],
    list[int],
    list[int],
    set[int],
    list[tuple[int, int]],
]


def build_initial_state(target: int, size: int, known_grid: Optional[Grid], game_mode: GameMode) -> InitialState:
    if size < 2:
        raise ValueError("size must be at least 2")
    if target <= size:
        raise ValueError("target must be greater than size")

    max_cell_value = resolve_max_cell_value(target=target, size=size, game_mode=game_mode)
    if max_cell_value < MIN_VALUE:
        raise ValueError("No valid value range is available for this game mode and target")

    grid = validate_and_normalize_known_grid(size=size, known_grid=known_grid, max_cell_value=max_cell_value)

    row_sums = [0] * size
    col_sums = [0] * size
    row_unknowns = [0] * size
    col_unknowns = [0] * size
    diag_sums = [0, 0]
    diag_unknowns = [0, 0]
    used_values: set[int] = set()
    unknown_positions: list[tuple[int, int]] = []

    for r in range(size):
        for c in range(size):
            value = grid[r][c]
            if value is None:
                row_unknowns[r] += 1
                col_unknowns[c] += 1
                for diag_index in diag_indexes(r, c, size):
                    diag_unknowns[diag_index] += 1
                unknown_positions.append((r, c))
            else:
                if value in used_values:
                    raise ValueError("known_grid cannot contain duplicate values")
                used_values.add(value)
                row_sums[r] += value
                col_sums[c] += value
                for diag_index in diag_indexes(r, c, size):
                    diag_sums[diag_index] += value

    validate_partial_sums(target, size, row_sums, row_unknowns, "row")
    validate_partial_sums(target, size, col_sums, col_unknowns, "column")
    validate_diagonal_partial_sums(target, diag_sums, diag_unknowns)

    return (
        grid,
        row_sums,
        col_sums,
        row_unknowns,
        col_unknowns,
        diag_sums,
        diag_unknowns,
        used_values,
        unknown_positions,
    )


def apply_value(
    value: int,
    r: int,
    c: int,
    size: int,
    grid: Grid,
    row_sums: list[int],
    col_sums: list[int],
    row_unknowns: list[int],
    col_unknowns: list[int],
    diag_sums: list[int],
    diag_unknowns: list[int],
    used_values: set[int],
) -> None:
    grid[r][c] = value
    row_sums[r] += value
    col_sums[c] += value
    row_unknowns[r] -= 1
    col_unknowns[c] -= 1
    for diag_index in diag_indexes(r, c, size):
        diag_sums[diag_index] += value
        diag_unknowns[diag_index] -= 1
    used_values.add(value)


def revert_value(
    value: int,
    r: int,
    c: int,
    size: int,
    grid: Grid,
    row_sums: list[int],
    col_sums: list[int],
    row_unknowns: list[int],
    col_unknowns: list[int],
    diag_sums: list[int],
    diag_unknowns: list[int],
    used_values: set[int],
) -> None:
    grid[r][c] = None
    row_sums[r] -= value
    col_sums[c] -= value
    row_unknowns[r] += 1
    col_unknowns[c] += 1
    for diag_index in diag_indexes(r, c, size):
        diag_sums[diag_index] -= value
        diag_unknowns[diag_index] += 1
    used_values.remove(value)


def final_constraints_met(target: int, size: int, row_sums: list[int], col_sums: list[int], diag_sums: list[int]) -> bool:
    return (
        all(row_sums[i] == target for i in range(size))
        and all(col_sums[i] == target for i in range(size))
        and diag_sums[0] == target
        and diag_sums[1] == target
    )

File: solvers/go_solver/state.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func buildRemainingAvailable(maxVal int, used map[int]bool, exclude int) []int {
	values := []int{}
	for v := 1; v <= maxVal; v++ {
		if used[v] || v == exclude {
			continue
		}
		values = append(values, v)
	}
	return values
}

func applyValue(st *State, value, r, c int) {
	st.grid[r][c] = value
	st.rowSums[r] += value
	st.colSums[c] += value
	st.rowUnknowns[r]--
	st.colUnknowns[c]--
	for _, d := range diagIndexes(r, c, st.size) {
		st.diagSums[d] += value
		st.diagUnknowns[d]--
	}
	st.used[value] = true
}

func revertValue(st *State, value, r, c int) {
	st.grid[r][c] = 0
	st.rowSums[r] -= value
	st.colSums[c] -= value
	st.rowUnknowns[r]++
	st.colUnknowns[c]++
	for _, d := range diagIndexes(r, c, st.size) {
		st.diagSums[d] -= value
		st.diagUnknowns[d]++
	}
	delete(st.used, value)
}

func buildState(req Request) (*State, error) {
	if req.Size < 2 {
		return nil, fmt.Errorf("size must be at least 2")
	}
	if req.Target <= req.Size {
		return nil, fmt.Errorf("target must be greater than size")
	}
	if req.GameMode == "" {
		req.GameMode = "unbounded"
	}
	maxVal := req.Target - 1
	if req.GameMode == "bounded_by_size_squared" {
		maxVal = min(maxVal, req.Size*req.Size)
	} else if req.GameMode != "unbounded" {
		return nil, fmt.Errorf("game_mode must be one of: unbounded, bounded_by_size_squared")
	}
	if maxVal < 1 {
		return nil, fmt.Errorf("no valid value range")
	}

	grid := make([][]int, req.Size)
	for i := 0; i < req.Size; i++ {
		grid[i] = make([]int, req.Size)
	}
	if req.KnownGrid != nil {
		if len(req.KnownGrid) != req.Size {
			return nil, fmt.Errorf("known_grid must match size")
		}
		for r := 0; r < req.Size; r++ {
			if len(req.KnownGrid[r]) != req.Size {
				return nil, fmt.Errorf("known_grid must match size")
			}
			for c := 0; c < req.Size; c++ {
				if req.KnownGrid[r][c] == nil {
					continue
				}
				value := *req.KnownGrid[r][c]
				if value < 1 || value > maxVal {
					return nil, fmt.Errorf("known value out of range")
				}
				grid[r][c] = value
			}
		}
	}

	st := &State{
		grid:        grid,
		target:      req.Target,
		size:        req.Size,
		maxVal:      maxVal,
		rowSums:     make([]int, req.Size),
		colSums:     make([]int, req.Size),
		rowUnknowns: make([]int, req.Size),
		colUnknowns: make([]int, req.Size),
		used:        map[int]bool{},
		rng:         rand.New(rand.NewSource(time.Now().UnixNano())),
	}

	for r := 0; r < req.Size; r++ {
		for c := 0; c < req.Size; c++ {
			value := st.grid[r][c]
			if value == 0 {
				st.rowUnknowns[r]++
				st.colUnknowns[c]++
				for _, d := range diagIndexes(r, c, st.size) {
					st.diagUnknowns[d]++
				}
				st.unknowns = append(st.unknowns, [2]int{r, c})
				continue
			}
			if st.used[value] {
				return nil, fmt.Errorf("known_grid cannot contain duplicate values")
			}
			st.used[value] = true
			st.rowSums[r] += value
			st.colSums[c] += value
			for _, d := range diagIndexes(r, c, st.size) {
				st.diagSums[d] += value
			}
		}
	}

	return st, nil
}
